examples/Example3/init.cc

```diff
@@ -0,0 +1,55 @@
+// SPDX-FileCopyrightText: 2020 CERN
+// SPDX-License-Identifier: Apache-2.0
+
+#include "init.h"
+
+#include "particle.h"
+#include "random.h"
+#include "user.h"
+
+static __host__ __device__ void init_particle(particle &p)
+{
+	p.time = 0.0f;
+	p.mass = uniform(1.0f, 10.0f);
+	p.charge = uniform() > 0.3f ? 0.0f : 1.0f;
+	p.position = {0.0f, 0.0f, 0.0f};
+	p.velocity = {100.0f, 0.0f, 0.0f};
+	p.acceleration = {0.0f, 0.0f, 0.0f};
+	p.alive = true;
+}
+
+#ifdef __CUDACC__
+static __global__ void init_gpu(int nparticles)
+{
+	for(unsigned int id = blockIdx.x * blockDim.x + threadIdx.x;
+		id < nparticles; id += blockDim.x * gridDim.x) {
+		particle p;
+		init_particle(p);
+		track_init(p);
+		set_particle(id, p);
+	}
+}
+#endif
+
+void init(int nparticles)
+{
+#ifndef __CUDACC__
+	particle p;
+	while (nparticles-- > 0) {
+		init_particle(p);
+		create_particle(p);
+	}
+#else
+	/* allocate and initialize rng states */
+	rng_alloc(nparticles);
+
+	rng_init<<<nparticles / 36 + 1, 32>>>(nparticles, 1);
```

Comment:
I did not realize it was a grid-size loop, so it is fine. However, then why picking nparticles/36 blocks and not something independent of nparticles?

Reply:
36 is the number of SMs in my GPU, so I divided the work in large enough blocks to have one block per SM to avoid scheduling overhead. I measured performance and this was quite a good choice compared with other options, such as choosing a fixed size for the blocks. Of course that at this point this is not such an important choice, but as I add documentation and comment the code better, these choices will become clearer. In the real simulation code, we'd probably need to query the number of SMs and use that to set an appropriate block size.